// @urql/core/dist/06b695c4.mjs
function h() {
  return (h = Object.assign || function(a) {
    var b, c, d;
    for (b = 1; b < arguments.length; b++) {
      c = arguments[b];
      for (d in c) {
        Object.prototype.hasOwnProperty.call(c, d) && (a[d] = c[d]);
      }
    }
    return a;
  }).apply(this, arguments);
}

function l(a) {
  return "string" == typeof a ? new GraphQLError(a) : "object" == typeof a && a.message ? new GraphQLError(a.message, a.nodes, a.source, a.positions, a.path, a, a.extensions || {}) : a;
}

function m() {
  return this.message;
}

function q(a, b) {
  a |= 0;
  for (var c = 0, d = 0 | b.length; c < d; c++) {
    a = (a << 5) + a + b.charCodeAt(c);
  }
  return a;
}

function u(a) {
  var b, c, d, e, f, g;
  if (null === a || r.has(a)) {
    return "null";
  }
  if ("object" != typeof a) {
    return JSON.stringify(a) || "";
  }
  if (a.toJSON) {
    return u(a.toJSON());
  }
  if (Array.isArray(a)) {
    for (b = "[", c = 0, d = a.length; c < d; c++) {
      0 < c && (b += ",");
      b += 0 < (e = u(a[c])).length ? e : "null";
    }
    return b + "]";
  }
  if (!(b = Object.keys(a).sort()).length && a.constructor && a.constructor !== Object) {
    return b = t.get(a) || Math.random().toString(36).slice(2), t.set(a, b), '{"__key":"' + b + '"}';
  }
  r.add(a);
  c = "{";
  d = 0;
  for (e = b.length; d < e; d++) {
    (g = u(a[f = b[d]])) && (1 < c.length && (c += ","), c += u(f) + ":" + g);
  }
  r.delete(a);
  return c + "}";
}

function v(a) {
  r.clear();
  return u(a);
}

function w(a) {
  return q(5381, a.replace(/([\s,]|#[^\n\r]+)+/g, " ").trim()) >>> 0;
}

function y(a) {
  var b, c, d;
  for (b = 0, c = a.definitions.length; b < c; b++) {
    if ((d = a.definitions[b]).kind === Kind.OPERATION_DEFINITION && d.name) {
      return d.name.value;
    }
  }
}

function z(a, b, c) {
  return {
    operation: a,
    data: b.data,
    error: Array.isArray(b.errors) ? new n({
      graphQLErrors: b.errors,
      response: c
    }) : void 0,
    extensions: "object" == typeof b.extensions && b.extensions || void 0
  };
}

function A(a, b, c) {
  return {
    operation: a,
    data: void 0,
    error: new n({
      networkError: b,
      response: c
    }),
    extensions: void 0
  };
}

function makeFetchBody(a) {
  return {
    query: print(a.query),
    operationName: y(a.query),
    variables: a.variables || void 0,
    extensions: void 0
  };
}

function makeFetchURL(a, b) {
  var c = a.context.url;
  if ("query" !== a.kind || !a.context.preferGetMethod || !b) {
    return c;
  }
  a = [];
  b.operationName && a.push("operationName=" + encodeURIComponent(b.operationName));
  b.query && a.push("query=" + encodeURIComponent(b.query.replace(/([\s,]|#[^\n\r]+)+/g, " ").trim()));
  b.variables && a.push("variables=" + encodeURIComponent(v(b.variables)));
  b.extensions && a.push("extensions=" + encodeURIComponent(v(b.extensions)));
  return c + "?" + a.join("&");
}

function makeFetchOptions(a, b) {
  var c = "query" === a.kind && !!a.context.preferGetMethod;
  return h({}, a = "function" == typeof a.context.fetchOptions ? a.context.fetchOptions() : a.context.fetchOptions || {}, {
    body: !c && b ? JSON.stringify(b) : void 0,
    method: c ? "GET" : "POST",
    headers: c ? a.headers : h({}, {
      "content-type": "application/json"
    }, a.headers)
  });
}

function makeFetchSource(a, b, c) {
  return make((function(d) {
    var e = d.next, f = d.complete, g = "undefined" != typeof AbortController ? new AbortController : null, p = !1;
    Promise.resolve().then((function() {
      if (!p) {
        return g && (c.signal = g.signal), function B(a, b, c) {
          var e, d = !1;
          return (a.context.fetch || fetch)(b, c).then((function(a) {
            e = a;
            d = 200 > a.status || a.status >= ("manual" === c.redirect ? 400 : 300);
            return a.json();
          })).then((function(b) {
            if (!("data" in b) && !("errors" in b)) {
              throw Error("No Content");
            }
            return z(a, b, e);
          })).catch((function(b) {
            if ("AbortError" !== b.name) {
              return A(a, d ? Error(e.statusText) : b, e);
            }
          }));
        }(a, b, c);
      }
    })).then((function(a) {
      p || (p = !0, a && e(a), f());
    }));
    return function() {
      p = !0;
      g && g.abort();
    };
  }));
}

function createRequest(a, b) {
  if ("string" == typeof a) {
    var c = w(a);
    a = void 0 !== x[c] ? x[c] : parse(a, {
      noLocation: !0
    });
  } else {
    void 0 !== a.__key ? c = a.__key : (c = w(print(a)), a = void 0 !== x[c] ? x[c] : a);
  }
  x[c] = a;
  a.__key = c;
  return {
    key: b ? q(c, v(b)) >>> 0 : c,
    query: a,
    variables: b || {}
  };
}

var n, r, t, x;

import { GraphQLError } from "graphql/error/GraphQLError.mjs";

import { parse } from "graphql/language/parser.mjs";

import { print } from "graphql/language/printer.mjs";

import { Kind } from "graphql/language/kinds.mjs";

import { make } from "wonka";

n = function(a) {
  function b(b) {
    var f, c = b.networkError, e = b.response;
    f = function k(a, b) {
      var d = "";
      if (void 0 !== a) {
        return d = "[Network] " + a.message;
      }
      void 0 !== b && b.forEach((function c(a) {
        d += "[GraphQL] " + a.message + "\n";
      }));
      return d.trim();
    }(c, b = (b.graphQLErrors || []).map(l));
    a.call(this, f);
    this.name = "CombinedError";
    this.message = f;
    this.graphQLErrors = b;
    this.networkError = c;
    this.response = e;
  }
  a && (b.__proto__ = a);
  (b.prototype = Object.create(a && a.prototype)).constructor = b;
  b.prototype.toString = m;
  return b;
}(Error);

r = new Set, t = new WeakMap;

x = Object.create(null);

export { n as C, h as _, A as a, makeFetchBody as b, makeFetchURL as c, makeFetchOptions as d, makeFetchSource as e, createRequest as f, y as g, z as m, v as s };
